pkg/kubefed2/federate.go

```diff
@@ -0,0 +1,292 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubefed2
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/golang/glog"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+
+	apiextv1b1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	apiextv1b1client "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/typed/apiextensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	apicommon "github.com/kubernetes-sigs/federation-v2/pkg/apis/core/common"
+	"github.com/kubernetes-sigs/federation-v2/pkg/apis/core/typeconfig"
+	fedv1a1 "github.com/kubernetes-sigs/federation-v2/pkg/apis/core/v1alpha1"
+	controllerutil "github.com/kubernetes-sigs/federation-v2/pkg/controller/util"
+	"github.com/kubernetes-sigs/federation-v2/pkg/kubefed2/options"
+	"github.com/kubernetes-sigs/federation-v2/pkg/kubefed2/util"
+)
+
+var (
+	federate_long = `
+	Federate enables a Kubernetes API type (including a CRD) to be
+	propagated to members of a federation.  Federation primitives will be
+	generated as CRDs and a FederatedTypeConfig will be created to
+	configure a sync controller.
+
+    Current context is assumed to be a Kubernetes cluster
+    hosting the federation control plane. Please use the
+    --host-cluster-context flag otherwise.`
+
+	federate_example = `
+	# Enable federation of a Kubernetes type by specifying the
+	# Kind, version and group (optional for core types) of the
+	# target type. The context of the federation control plane's
+	# host cluster must be supplied if it is not the current
+	# context.
+	kubefed2 federate --group= --version=v1 --kind=Secret --host-cluster-context=bar`
```

Comment:
Should `federate` be able to support different versions of the same group/kind? See:
```
$ ./bin/kubefed2 federate --namespaced=false --group=autoscaling --version=v1 --kind=HorizontalPodAutoscaler
$ kubectl get crd | grep hor
federatedhorizontalpodautoscalerplacements.generated.federation.k8s.io   2018-10-19T18:02:52Z
federatedhorizontalpodautoscalers.generated.federation.k8s.io            2018-10-19T18:02:52Z
$ ./bin/kubefed2 federate --namespaced=false --group=autoscaling --version=v2beta1 --kind=HorizontalPodAutoscaler
F1019 18:04:24.862496    3656 federate.go:113] error: Error creating CRD "federatedhorizontalpodautoscalers.generated.federation.k8s.io": customresourcedefinitions.apiextensions.k8s.io "federatedhorizontalpodautoscalers.generated.federation.k8s.io" already exists
```

Reply:
OK. I'm trying to think through lifecycle management of federated resources using fed-v2. Here is the use case:

1. I have an older federated resource (i.e.  `FederatedHorizontalPodAutoscaler` v1beta1).
2. I need to upgrade the resource (i.e. v1, along with other attributes).
3. I want to deploy the new resource version while still running the old version.
4. I use the `FederatedHorizontalPodAutoscalerPlacement` to deploy the federated resource to a single cluster (test/dev). After all goes well, I update the `FederatedHorizontalPodAutoscalerPlacement` to include all clusters.
5. After the new version runs for a period of time, I remove the older version of `FederatedHorizontalPodAutoscaler`.

I think the above use case will be common. Without supporting different resource versions, I don't think fed-v2 can support this use case. Am I missing something or how would you go about supporting the above use case?

Reply:
If I create `FederatedHorizontalPodAutoscaler` `foo` in namespace `bar` with version `v1alpha`, trying to create that same resource with same name with version 'v1' will fail because the resource already exists.  That's because the version doesn't affect the storage of the resource, only how the API endpoint receives requests (e.g. allowing new fields or values at a newer endpoint).

If a member of the federation is upgraded to a newer version of Kubernetes that supports a newer version of `HorizontalPodAutoscaler`, propagating resources to that cluster automatically get  upgraded behavior regardless of the version of the API they are targeting.  Because of the unified storage and backwards compatibility guarantees, a `v1beta1` request to an API that supports `v1` will automatically apply defaults to added fields to ensure `v1` compatibility.  The HPA controller won't know what version of the API was used to create the resource.

That's not to say that the version of the target resource defined in the FederatedTypeConfig isn't important.  If a given version of HPA supports field `foo`, the version of the target resource must be set to that version or greater for the sync controller to be able to successfully deliver that field to a member cluster.

Reply:
Lets say the use case is: I have a 3 cluster federation. Each cluster runs k8s v1.11. I need to upgrade to 1.12. I start off upgrading my test/dev cluster. All goes well, but v1.12 upgrades resource X from `v1beta2` to `v1`. I need fed-v2 to support both of these versions so resources can be managed across 1.11 and 1.12 clusters until all 3 clusters are upgraded to 1.12.

Reply:
BTW, thanks for all the feedback above.

Reply:
Because of the backwards compatibility guarantees that Kubernetes provides for APIs, upgrading a member cluster to 1.12 would enable `v1` for type X while retaining support for `v1beta2`.  Since the sync controller for type X would need to target the version common to all members of the federation, the controller would need to continue targeting `v1beta2` until all clusters were updated to support `v1`.